src/persist-client/examples/persistent_source.rs

```diff
@@ -0,0 +1,568 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! An example that shows how a highly-available (HA) persistent source could be implemented on top
+//! of the new persist API, along with a way to do compare-and-set updates of a collection of
+//! timestamp bindings. In our case, what we need to make the source HA is that multiple instances
+//! can run concurrently and produce identical data. We achieve this by making the minting of new
+//! timestamp bindings "go through" consensus. In our case, some system that supports a
+//! compare-and-set operation.
+//!
+//! The example consists of three parts, an [`api`] module, an [`impls`] module, and a [`render`]
+//! module. The `api` module has traits for defining sources and a timestamper, `impls` has example
+//! implementations of these traits, and [`render`] can spawn (or "render") the machinery to read
+//! from a source and write it into persistence.
+
+// TODO(aljoscha): Actually write to persistence and use persistence (or another consensus system)
+// to mint timestamp bindings.
+// TODO(aljoscha): Figure out which of the Send + Sync shenanigans are needed and/or if the
+// situation can be improved.
+
+use std::error::Error;
+use std::time::Duration;
+
+use tracing::error;
+
+use mz_ore::now::SYSTEM_TIME;
+
+use crate::impls::SequentialSource;
+use crate::impls::SharedTimestamper;
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn Error>> {
```

Comment:
I came to regret my suggestion to structure this as a "cargo example" last time since that means it's a separate binary to be linked. if it's not too much trouble, how about making this a `#[test]` of some sort even if it doesn't really have assertions. and then since the runtime will be long from the sleeps, let's toss on an `#[ignore]` too so it has to be specifically requested

Reply:
can do!